File: 230. Kth Smallest Element in a BST.cpp
#include<vector>
#include<iostream>


using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
    public:

        void printVector(vector<int> nums){
            for(int i=0;i<nums.size();i++){
                cout << " " << nums[i] << " ";
            }
            cout << endl;
        }

        void depthFirstTraversal(TreeNode* node){
            
            if(!node){
                return;
            }

            // init empty heap & do the work
            treeValues.push_back(node->val);

            depthFirstTraversal(node->left);
            depthFirstTraversal(node->right);


        }

        int kthSmallest(TreeNode* root, int k) {
            
            depthFirstTraversal(root);

            sort(treeValues.begin(),treeValues.end());

            for(int i=0;i<treeValues.size();i++){
                if(i==k-1){
                    return treeValues[i];
                }
            }

            return -1;
        }

    private: 
        vector<int> treeValues;
        int depth = 0;
};

int main(){

    TreeNode* root = new TreeNode(3);
    root->right = new TreeNode(4);
    TreeNode* n2 = new TreeNode(1);
    n2->right = new TreeNode(2);
    root->left = n2;

    Solution solution;

    int answer = solution.kthSmallest(root,1);

    cout << "Answer is: " << answer << endl;





    return 0;
}

File: 230. Kth Smallest Element in a BST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        
        arr = []

        # L - N - R
        def dfs(node):
            if node == None: return
            dfs(node.left)
            arr.append(node.val)
            dfs(node.right)
        
        dfs(root)
        return arr[k-1]

    def kthSmallestIterative(self, root: Optional[TreeNode], k: int) -> int:
        
        s = collections.deque()
        cur = root

        while s or cur:

            while cur:
                s.append(cur)
                cur = cur.left
            
            cur = s.pop()
            k -= 1
            if k == 0: return cur.val
            cur = cur.right
